proxy/http/parser/parser.py

```diff
@@ -300,19 +299,28 @@ def _process_line_and_headers(self, raw: bytes) -> Tuple[bool, bytes]:
         return len(raw) > 0, raw
 
     def _process_line(self, raw: bytes) -> None:
+        # defer import to solve the circular import problem
+        from ..exception import HttpProtocolException
         if self.type == httpParserTypes.REQUEST_PARSER:
-            # parse request line according https://datatracker.ietf.org/doc/html/rfc2616#section-5.1
+            # Ref:
+            # https://datatracker.ietf.org/doc/html/rfc2616#section-5.1
+            # https://greenbytes.de/tech/webdav/rfc7230.html#request.line
+            # https://greenbytes.de/tech/webdav/rfc7231.html#methods
+            # http://www.iana.org/assignments/http-methods/http-methods.xhtml
             if self.protocol is not None and self.protocol.version is None:
                 # We expect to receive entire proxy protocol v1 line
                 # in one network read and don't expect partial packets
                 self.protocol.parse(raw)
             else:
                 line = raw.split(WHITESPACE)
-                if len(line) == 3:
-                    self.method = line[0].upper()
+                if len(line) == 3 and line[0] in httpMethods:
```

Comment:
Thank you.  But additional check on `line[0] in httpMethods` must be removed.  Reason being, `HttpParser` is common for web and proxy server components.  While this check is valid for a proxy server, a web server can use custom HTTP method verbs, which are valid as per specification.

If a parsed method verb is invalid, then it should not be a parser error, but actual error response must be returned from server handlers. E.g. returning `Method not supported error`.  For reference see screenshot below from linked RFC

<img width="599" alt="Screen Shot 2021-11-15 at 7 33 00 PM" src="https://user-images.githubusercontent.com/126065/141795182-833fb7a9-45ca-480c-bb8e-974738a8d27b.png">

Reply:
Maybe we should update the supported method according to 
https://www.iana.org/assignments/http-methods/http-methods.xhtml

Reply:
I will remove the method check for now

Reply:
Yes lets add those methods within `httpMethods` named tuple.  About restricting verbs to only defined verb, I am not sure about it still.  E.g. why stop someone from using a custom verb in first place.  How does it hurt.  In the end the server will anyways return unsupported method if received verb is invalid within its context.  Hence I think we must allow parsing custom verbs and let server handlers do their job or rejecting the request.

Wdyt?

Reply:
We can remove the method restriction for now and let server handlers to do there job, that will give more flexibility